components/support/error/src/handling.rs

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Helpers for components to "handle" errors.
+
+/// Describes what error reporting action should be taken.
+pub enum ErrorReporting {
+    // No logging or error reporting.
+    Nothing,
+    // We write a log message but don't report it.
+    Log {
+        level: log::Level,
+    },
+    // We log a message and report via our error reporter.
+    Report {
+        level: log::Level,
+        report_class: String,
+    },
+}
+
+// Specifies how an "internal" error is converted to an "external" public error and
+// any logging or reporting that should happen.
+pub struct ErrorHandling<E> {
+    // The external error that should be returned.
+    pub err: E,
+    // How the error should be reported.
+    pub reporting: ErrorReporting,
+}
+
+pub trait GetErrorHandling {
+    type ExternalError;
+
+    // Return how to handle our internal errors
+    fn get_error_handling(&self) -> ErrorHandling<Self::ExternalError>;
+
+    // Some helpers to cut the verbosity down.
+    // Just convert the error without any special logging or error reporting.
+    fn passthrough(err: Self::ExternalError) -> ErrorHandling<Self::ExternalError> {
```

Comment:
Would implementers of the trait ever override this? _maybe_ can get rid of this whole trait and replace it with a bunch of `fn passthrough<E>(err: E) -> ErrorHandling<E>` etc functions and get rid of the trait?

i.e tell consumers to implement their own `impl<IE, EE> From<IE> for ErrorHandling<EE>` etc

Not sure how good of an idea that is, and this setup maybe better 🤷 but the thought crossed my mind so sharing it

Reply:
I was thinking some crates might want to override some of them maybe, but haven't really thought it through - eg, I could imagine one crate overriding `passthrough` to also log at a trace level or something like that - but then again, maybe that's not the right way to approach this. Maybe these "helpers" don't actually help and are just a potentially confusing extra level of indirection that should just be removed so things are more explicit and less magic?

It's a good thing for us to consider and I don't really have a strong sense of what's better here. We could also consider converting one or 2 other crates to reuse this mechanism too - fxaccounts would be a good candidate - to help get a better feel for it?

Reply:
I personally want to go the other way and shave off a level of indirection.  If all these functions are doing is providing a convenient way to construct an `ErrorHandling` instance, then they should be in `ErrorHandling` rather than the trait.